core/src/Cuda/Kokkos_CudaSpace.cpp

```diff
@@ -172,6 +174,115 @@ void *CudaSpace::allocate(const char *arg_label, const size_t arg_alloc_size,
 }
 
 namespace {
+
+#ifndef CUDART_VERSION
+#error CUDART_VERSION undefined!
+#elif (defined(KOKKOS_ENABLE_IMPL_CUDA_MALLOC_ASYNC) && CUDART_VERSION >= 11020)
+
+/**
+ *  If a device has been initialized we will drop it in the set
+ * so we don't initialize it again
+ */
+std::unordered_set<int> _mem_pool_initialized;
+
+/**
+ * This is the initializer. It currently gets the device_id and stream
+ * that impl_alloc gets
+ *    @param device_id  The device id on which to alter the default mempool
+ *    @param stream     A specific stream for cudaMallocAsync
+ * 		@param error_code return any failure codes for cuda APIs
+ *    @return  				  true for success, false for failure
+ */
+bool initializeMempool(const int device_id, const cudaStream_t stream,
+                       cudaError_t &error_code) {
+  error_code = cudaSuccess;
+
+  std::cout << "Initializing Default Memory Pool for device " << device_id
+            << "\n";
+
+  // Interpret env var
+  char *env_string = getenv("KOKKOS_CUDA_MEMPOOL_SIZE");
+  if (!env_string) return true;  // Nothing to set, initialization is complete
+
+  // Env var can be a decimal e.g. 2.4g
+  double factor = 1;  // bytes
+
+  // Interpret the string
+  std::string mempool_size_string(env_string);
+  char last_letter = mempool_size_string.back();
+
+  // We can deal with the last letter being units.
+  if (std::isalpha(last_letter)) {
+    mempool_size_string.pop_back();
+    char last_letter_cap = std::toupper(last_letter);
+    switch (last_letter_cap) {
+      case 'K': factor = 1024; break;
+      case 'M': factor = 1024 * 1024; break;
+      case 'G': factor = 1024 * 1024 * 1024; break;
+      default:
+        // Not permitted
+        return false;
```

Comment:
This should probably print the problem.
```suggestion
        std::cerr << "KOKKOS_CUDA_MEMPOOL_SIZE couldn't be parsed properly!\n"
        // Not permitted
        return false;
```

Reply:
So I can change it to that. Currently what will happen is that false will be returned but the `error_code` will still be `cudaSuccess` and the error is dealt with as an exception on line 320. The logic is that if this routine returns false either a CUDA API failed or the parsing failed. The failure of the CUDA API can be determined by looking at `error_code`

However I am happy to accommodate whichever way you prefer to promulgate the error to the user. I can put the error here, or at the point of raising the exception.

Reply:
Added a brief report here about not being able to parse units in af41845ef